core/app/models/spree/wallet_payment_source.rb

```diff
@@ -1,19 +1,35 @@
 # frozen_string_literal: true
 
-class Spree::WalletPaymentSource < ActiveRecord::Base
-  belongs_to :user, class_name: Spree::UserClassHandle.new, foreign_key: 'user_id', inverse_of: :wallet_payment_sources
-  belongs_to :payment_source, polymorphic: true, inverse_of: :wallet_payment_sources
+module Spree
+  class WalletPaymentSource < Spree::Base
+    belongs_to :user, class_name: Spree::UserClassHandle.new, foreign_key: 'user_id', inverse_of: :wallet_payment_sources
+    belongs_to :payment_source, polymorphic: true, inverse_of: :wallet_payment_sources
 
-  validates_presence_of :user
-  validates_presence_of :payment_source
+    validates_presence_of :user
+    validates_presence_of :payment_source
+    validates :user_id, uniqueness: {
+      scope: [:payment_source_type, :payment_source_id],
+      message: :payment_source_already_exists
```

Comment:
This could potentially create issues with payment sources that have already been added to a user and are no more valid - for example, if you already have duplicates and try to update one of them, right?

I'm not sure what is the best way to handle this, maybe use this validation only [on create](https://guides.rubyonrails.org/active_record_validations.html#on) so it does not apply to already added payment sources?

Also, how did you have this issue? Can you share the reason/scenario behind this problem (which I see exists)?

Reply:
just read https://github.com/solidusio/solidus/pull/3007#issuecomment-448727125 - nevermind